Filter empty samples after removing metadata from 20 newsgroup dataset

hugoabonizio opened the pull request:
I was trying out some models on the 20 newsgroup dataset by using the `fetch_20newsgroups` method, but was geting some blank samples, which was causing my preprocessing and models to fail. Then I figured out that the `remove` option may remove all the content of some samples, as can be seen bellow. So I'm proposing to filter out those empty samples after removing the metadata to avoid silently generate blank samples.

```python
from sklearn.datasets import fetch_20newsgroups

full = fetch_20newsgroups(subset='train')
removed = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'))

print(len(full.data[155])) # => 849
print(len(removed.data[155])) # => 0
```

hugoabonizio committed:
Filter empty samples after removing metadata from 20 newsgroup dataset

hugoabonizio committed:
Fix linting issue

glemaitre commented on sklearn/datasets/_twenty_newsgroups.py:
```diff
@@ -284,6 +284,18 @@ def fetch_20newsgroups(data_home=None, subset='train', categories=None,
     if 'quotes' in remove:
         data.data = [strip_newsgroup_quoting(text) for text in data.data]
 
+    if len(remove) > 0:
```
Would it be possible to make this selection in the previous filtering? Something along these lines:

```python
    strip_funcs = {
        "headers": strip_newsgroup_header,
        "footers": strip_newsgroup_footer,
        "quotes": strip_newsgroup_quoting,
    }

    for strip_type in remove:
        tmp_data = []
        for text in data:
            stripped_text = strip_funcs[strip_type](text)
            if not stripped_text:
                tmp_data.append(stripped_text)
        data.data = tmp_data
```

glemaitre reviewed (commented):
We will need a non-regressiont test for this case and an entry in what's new.

hugoabonizio committed:
Refact empty 20 newsgroups dataset samples filtering

hugoabonizio commented:
Sorry for the late reply @glemaitre. I introduced a `clear_empty` flag to avoid breaking compatibility with current working code and added an entry in what's new. I also took your suggestion and reworked the selection. Let me know what you think.
